contrib/core-contract-tests/tests/pox-4/pox-4.prop.test.ts

```diff
@@ -0,0 +1,1605 @@
+      );
+    });
+
+    it("should return none get-check-delegation", () => {
+      fc.assert(
+        fc.property(
+          fc.constantFrom(...simnet.getAccounts().values()),
+          (caller) => {
+            // Arrange
+            // Act
+            const { result: actual } = simnet.callReadOnlyFn(
+              "pox-4",
+              "get-check-delegation",
+              [Cl.principal(caller)],
+              caller
+            );
+            // Assert
+            assert(isClarityType(actual, ClarityType.OptionalNone));
+          }
+        )
+      );
+    });
+
+    it("should return none get-delegation-info", () => {
+      fc.assert(
+        fc.property(
+          fc.constantFrom(...simnet.getAccounts().values()),
+          (caller) => {
+            // Arrange
+            // Act
+            const { result: actual } = simnet.callReadOnlyFn(
+              "pox-4",
+              "get-delegation-info",
+              [Cl.principal(caller)],
+              caller
+            );
+            // Assert
+            assert(isClarityType(actual, ClarityType.OptionalNone));
+          }
+        )
+      );
+    });
+
+    it("should return correct get-pox-info", () => {
+      fc.assert(
+        fc.property(
+          fc.constantFrom(...simnet.getAccounts().values()),
+          (caller) => {
+            // Arrange
+            const expected_reward_cycle_id = 0,
+              expected_first_burn_block_height = 0;
+            // Act
+            const { result: actual } = simnet.callReadOnlyFn(
+              "pox-4",
+              "get-pox-info",
+              [],
+              caller
+            );
+            // Assert
+            assert(isClarityType(actual, ClarityType.ResponseOk));
+            assert(isClarityType(actual.value, ClarityType.Tuple));
+            expect(actual.value.data["first-burnchain-block-height"]).toBeUint(
+              expected_first_burn_block_height
+            );
+            expect(actual.value.data["min-amount-ustx"]).toBeUint(
+              MIN_AMOUNT_USTX
+            );
+            expect(actual.value.data["prepare-cycle-length"]).toBeUint(
+              TESTNET_PREPARE_CYCLE_LENGTH
+            );
+            expect(actual.value.data["reward-cycle-id"]).toBeUint(
+              expected_reward_cycle_id
+            );
+            expect(actual.value.data["reward-cycle-length"]).toBeUint(
+              TESTNET_REWARD_CYCLE_LENGTH
+            );
+            expect(actual.value.data["total-liquid-supply-ustx"]).toBeUint(
+              INITIAL_TOTAL_LIQ_SUPPLY
+            );
+          }
+        )
+      );
+    });
+
+    it("should return none get-allowance-contract-caller", () => {
+      fc.assert(
+        fc.property(
+          fc.constantFrom(...simnet.getAccounts().values()),
+          fc.constantFrom(...simnet.getAccounts().values()),
+          fc.constantFrom(...simnet.getAccounts().values()),
+          (caller, sender, contract_caller) => {
```

Comment:
What happens if any of the caller, sender, and contract_caller are the same?

Reply:
Nothing, it just does not make sense that the `sender` and the `contract_caller` are the same in the context of pox-4 allowance. The `allow-contract-caller` method is called to allow a pool contract or another principal to stack on behalf of a principal.